bundler/lib/dependabot/bundler/native_helpers.rb

```diff
@@ -49,8 +49,7 @@ def self.run_bundler_subprocess(function:, args:, bundler_version:, options: {})
               # Bundler will pick the matching installed major version
               "BUNDLER_VERSION" => bundler_version,
               "BUNDLE_GEMFILE" => File.join(helpers_path, "Gemfile"),
-              # Prevent the GEM_HOME from being set to a folder owned by root
-              "GEM_HOME" => File.join(helpers_path, ".bundle")
+              "GEM_PATH" => File.join(helpers_path, "vendor/cache")
```

Comment:
`GEM_PATH` defines where gems are loaded, whereas `GEM_HOME` has to do with where gems are installed by default. We set `GEM_PATH` here to point to a directory created by `bundle package`, which creates a self-contained distribution of vendored gems.

Reply:
The `vendor/cache` folder generated by `bundle package` is not a proper `GEM_PATH` or `GEM_HOME` location, it's just a folder with `.gem` packages inside to act as a local cache of gems for an application, so that `bundle install` doesn't need to use the network for fetching `.gem` packages.

I don't think the current setup, where `GEM_HOME` is set to a `.bundle` folder works as expected either, because `bundle config --local path .bundle` does not actually install gems to `.bundle`, but to `.bundle/ruby/2.7.0` instead (this is so that switching rubies does not reuse the installation path and mess up your native gems that will have different byte code depending on the Ruby version).

So basically the current setup is essentially setting `GEM_HOME` to an empty folder. I think the reason things still work currently is because `GEM_PATH` is unchanged and still includes the default `GEM_HOME` which has the proper version of bundler installed.

To provide a self-contained distribution of gems one normally uses `bundle config path`, which is what the code was previously doing, but I don't think it was working as expected anyways because:

* `bundle config --local path .bundle` does not actually set gems to be installed to `.bundle`, but to `.bundle/ruby/2.7.0` instead (this is so that switching rubies does not reuse the installation path and mess up your native gems that will have different byte code depending on the Ruby version).
* Even if that used a correct path, `bundle install` does not vendor the version of Bundler itself into the configured path. I think this is something we should improve upstream, but not there yet.

I created a patch the fixed tests for me locally, but it's just a WIP, not really happy with it yet. I'm working on simplifying things at #4973, but I share it here just in case:

```diff
diff --git a/bundler/Earthfile b/bundler/Earthfile
index 9c868ecc1..716943727 100644
--- a/bundler/Earthfile
+++ b/bundler/Earthfile
@@ -52,15 +52,12 @@ bundler-v1:
     COPY helpers/v1/Gemfile .
     COPY helpers/v1/run.rb .
 
-    RUN gem update --system $RUBYGEMS_SYSTEM_VERSION --no-document && \
-        gem install bundler:$BUNDLER_VERSION --no-document && \
-        bundle config --local without "test" && \
-        bundle install --jobs 4 --retry 3 && \
-        bundle package && \
-        rm -rf /var/lib/gems/*/cache/*
-
-    RUN cd vendor/cache && \
-        gem fetch bundler -v $BUNDLER_VERSION
+    RUN gem update --system $RUBYGEMS_SYSTEM_VERSION --no-document \
+     && gem install bundler:$BUNDLER_VERSION --no-document \
+     && bundle config --local path ".bundle" \
+     && bundle install --jobs 4 --retry 3 \
+     && gem install bundler:$BUNDLER_VERSION --no-document --install-dir .bundle/ruby/2.7.0 \
+     && rm -rf /var/lib/gems/*/cache/*
 
     SAVE ARTIFACT . helpers
 
@@ -74,14 +71,11 @@ bundler-v2:
 
     RUN gem update --system $RUBYGEMS_SYSTEM_VERSION --no-document \
      && gem install bundler:$BUNDLER_VERSION --no-document \
-     && bundle config --local without "test" \
+     && bundle config --local path ".bundle" \
      && bundle install --jobs 4 --retry 3 \
-     && bundle package \
+     && gem install bundler:$BUNDLER_VERSION --no-document --install-dir .bundle/ruby/2.7.0 \
      && rm -rf /var/lib/gems/*/cache/*
 
-    RUN cd vendor/cache \
-     && gem fetch bundler -v $BUNDLER_VERSION
-
     SAVE ARTIFACT . helpers
 
 deps:
diff --git a/bundler/lib/dependabot/bundler/native_helpers.rb b/bundler/lib/dependabot/bundler/native_helpers.rb
index d15f59907..b40783691 100644
--- a/bundler/lib/dependabot/bundler/native_helpers.rb
+++ b/bundler/lib/dependabot/bundler/native_helpers.rb
@@ -49,7 +49,8 @@ module Dependabot
               # Bundler will pick the matching installed major version
               "BUNDLER_VERSION" => bundler_version,
               "BUNDLE_GEMFILE" => File.join(helpers_path, "Gemfile"),
-              "GEM_PATH" => File.join(helpers_path, "vendor/cache")
+              # Prevent the GEM_HOME from being set to a folder owned by root
+              "GEM_HOME" => File.join(helpers_path, ".bundle", "ruby", "2.7.0")
             }
           )
         rescue SharedHelpers::HelperSubprocessFailed => e
```

Reply:
@deivid-rodriguez Thanks for explaining all that. A few follow-up questions:

Is `vendor` a valid `GEM_PATH`? My understanding is that RubyGems looks for `.gem` files in a `cache` subdirectory of a gems dir. Is that sufficient? Or does it also need the sources in `gems/<gem_name>/`? 

Do you know if there's a way to prevent Bundler from hardcoding that `ruby/2.7.0` path?

Alternatively, do you think we could we sidestep this issue with [Traveling Ruby](https://github.com/phusion/traveling-ruby)?

Reply:
Hei @mattt, I just noticed that while trying to reorganize my previous comment I copied over some stuff and ended up repeating the same things again and again, sorry 😓.

Anyways, regarding your questions. A valid path for `GEM_PATH` have the following structure:

```
ls -la /Users/deivid/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0 
total 0
drwxr-xr-x   11 deivid  staff    352 28 abr 16:53 .
drwxr-xr-x    3 deivid  staff     96 15 abr 12:41 ..
drwxr-xr-x    4 deivid  staff    128 28 abr 10:39 bin
drwxr-xr-x    2 deivid  staff     64 15 abr 12:41 build_info
drwxr-xr-x    3 deivid  staff     96 20 abr 08:42 bundler
drwxr-xr-x  502 deivid  staff  16064 28 abr 16:53 cache
drwxr-xr-x    4 deivid  staff    128 15 abr 22:40 doc
drwxr-xr-x    3 deivid  staff     96 15 abr 12:41 extensions
drwxr-xr-x  564 deivid  staff  18048 28 abr 16:53 gems
drwxr-xr-x    4 deivid  staff    128 26 abr 13:51 plugins
drwxr-xr-x  491 deivid  staff  15712 28 abr 16:53 specifications
```

When you configure `bundle config path vendor`, then `bundle install` will generate a similar structure at the `vendor/ruby/2.7.0` folder.

> My understanding is that RubyGems looks for .gem files in a cache subdirectory of a gems dir. Is that sufficient?

Correct, `bundle cache` provides something similar but that's supposed to be application specific, rather than global for your Ruby installation.

> Or does it also need the sources in gems/<gem_name>/?

Overall, I think setting `GEM_PATH` like we are doing now shouldn't be necessary in general, since Dependabot does not really installs any gems, only "resolves" them. I think it's needed exclusively for some code related to vendoring gems that needs to download some gems, so dependabot ends up trying to write to the `GEM_PATH` cache of gems, which is by default not writable. So, yeah, I think only the cache should be needed for us.

> Do you know if there's a way to prevent Bundler from hardcoding that ruby/2.7.0 path?

That should be easy to do, "ruby" is `RUBY_ENGINE`, and "2.7.0" is the ABI version, which would be something like `RUBY_VERSION.sub(/\d+\z/, "0")`.

Regarding, "traveling ruby", I'm honestly not sure.

Reply:
> Do you know if there's a way to prevent Bundler from hardcoding that ruby/2.7.0 path?

Oh, I think I misunderstood this question. You're asking if there's a way for Bundler to install to `vendor` directly instead of `vendor/ruby/2.7.0`, right? If that's the case, no, there's no such option at the moment.